pandas/core/dtypes/cast.py

```diff
@@ -110,7 +110,9 @@ def trans(x):  # noqa
                     np.prod(result.shape)):
                 return result
 
-        if issubclass(dtype.type, np.floating):
+        # don't convert bool to float GH16875
+        if (issubclass(dtype.type, np.floating) and
+                not is_bool_dtype(result.dtype)):
             return result.astype(dtype)
         elif is_bool_dtype(dtype) or is_integer_dtype(dtype):
```

Comment:
Sorry but I'm a bit confused. Are you sugesting the following change?
```
        if is_bool_dtype(dtype) or is_integer_dtype(dtype):
        elif issubclass(dtype.type, np.floating) :
             return result.astype(dtype)
```
This won't work because the guard checks on the type of `result.dtype` parameter whereas the second if checks whether `dtype` parameter is of type bool/int.